packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -0,0 +1,330 @@
+ */
+function getLeafAllowedTypes(schema: TreeFieldStoredSchema): string {
+	let result = "";
+	const leafTypes = schema.types;
+
+	if (leafTypes === undefined) {
+		result = "any";
+	} else {
+		for (const type of leafTypes) {
+			result += `${type} | `;
+		}
+	}
+
+	// Slice the trailing ` | ` from the `result`.
+	result = `${result.slice(0, -3)}`;
+
+	return result;
+}
+
+/**
+ * Returns the schema & leaf value of the node with type {@link LeafNodeStoredSchema}.
+ */
+function visualizeLeafNodeStoredSchema(
+	tree: JsonableTree,
+	allowedTypes: string,
+): SharedTreeLeafNode {
+	return {
+		schema: {
+			schemaType: SharedTreeSchemaType.LeafNodeStoredSchema,
+			allowedTypes,
+		},
+		value: JSON.stringify(tree.value),
+		kind: VisualSharedTreeNodeKind.LeafNode,
+	};
+}
+
+/**
+ * Returns the schema & fields of the node with type {@link ObjectNodeStoredSchema}.
+ */
+function visualizeObjectNodeStoredSchema(
+	tree: JsonableTree,
+	schema: ObjectNodeStoredSchema,
+	contentSnapshot: SharedTreeContentSnapshot,
+): VisualSharedTreeNode {
+	const treeFields = tree.fields;
+
+	if (treeFields === undefined || Object.keys(treeFields).length === 0) {
+		return {
+			schema: {
+				name: tree.type,
+				schemaType: SharedTreeSchemaType.ObjectNodeStoredSchema,
+				allowedTypes: getObjectAllowedTypes(schema),
+			},
+			fields: {},
+			kind: VisualSharedTreeNodeKind.InternalNode,
+		};
+	}
+
+	const fields: Record<string | number, VisualSharedTreeNode> = {};
+
+	// {@link EmptyKey} indicates an array field (e.g., `schemabuilder.array()`).
+	// Hides level of indirection by omitting the empty key in the visual output.
+	if (
+		Object.keys(treeFields).length === 1 &&
+		Object.prototype.hasOwnProperty.call(treeFields, EmptyKey)
+	) {
+		const children = treeFields[EmptyKey];
+
+		for (let i = 0; i < children.length; i++) {
+			const childSchema = contentSnapshot.schema.nodeSchema.get(children[i].type);
+			// If the child node is a leaf node, get the allowed types from the parent schema before entering {@link visualizeSharedTreeNodeBySchema}.
+			let leafAllowedTypes: string | undefined;
+
+			// If the node within the array is a leaf node, get the allowed types from the parent schema.
+			if (childSchema instanceof LeafNodeStoredSchema) {
+				const leafSchema = schema.objectNodeFields.get(EmptyKey);
+
+				if (leafSchema === undefined) {
+					throw new TypeError("Leaf schema should not be undefined.");
+				}
+
+				leafAllowedTypes = getLeafAllowedTypes(leafSchema);
+			}
+
+			fields[i] = visualizeSharedTreeNodeBySchema(
+				children[i],
+				childSchema,
+				contentSnapshot,
+				leafAllowedTypes,
+			);
+		}
+	} else {
+		for (const [fieldKey, childField] of Object.entries(treeFields)) {
+			const childSchema = contentSnapshot.schema.nodeSchema.get(childField[0].type);
+			let leafAllowedTypes: string | undefined;
+
+			// If the child field is a leaf node, get the allowed types from the parent schema.
```

Comment:
We repeat this leaf node pattern in a few places. If possible, we should refactor this out into some reusable utility.